deepspeed/runtime/zero/stage_1_and_2.py

```diff
@@ -589,9 +605,12 @@ def _configure_moe_settings(self):
         assert self.expert_dp_process_group is not None, "Expert data parallel group should be configured with MoE"
         assert self.ep_process_group is not None, "Expert parallel group should be configured with MoE"
 
-    def _update_model_bit16_weights(self, group_index):
-        updated_params = self.unflatten(self.bit16_groups_flat[group_index],
-                                        self.round_robin_bit16_groups[group_index])
+    def _update_model_bit16_weights(self, group_index, use_cpu_data=False):
+        if use_cpu_data:
```

Comment:
It seems that `use_cpu_data==True` means to unflatten from flattened device buffer into cpu buffers, is that correct? 
I am concerned that unflattening into cpu buffers will break line 615, perhaps only for non-hpu devices. This is because `p.data` will be updated from device buffer to cpu buffer (`q.data`). Can you please clarify?

Reply:
No. see my (long) comment above. it uses only the shapes from the tensors list to know how to unflatten.

Reply:
You can ignore my question here. Your previous response already clarified my confusion. Thanks!